Implement Feature, Notifier and other improvements

squirrelsc opened the pull request:
See commits for details

squirrelsc committed:
remove unused from_value methods

As final implementation, no need to create capability from value. The
capability should be defined, if it's needed. If a capability is not
defined, a node uses default capability.

squirrelsc committed:
change feature in schema to str, not a class.

The feature class will be parsed in platform level, so it doesn't need
to know feature class in schema. the excluded feature set won't be
merged into included, after this change.

squirrelsc committed:
update comments in executable.py

squirrelsc committed:
improve log of search_space

squirrelsc committed:
node: remove Dict base to avoid confusing

The nodes is a list, which supports queried by name. It doesn't
implement some dict method. So remove dict to get it clearer.

squirrelsc committed:
Process: remove unused code.

squirrelsc committed:
add cat command to output sys files

squirrelsc committed:
Improve SetSpace for serialization

The original collection of SetSpace cannot be serialized, so that azure
location cache cannot save features to disk. So add items field for
serialization.

Note, it's a minimum patch to support our functionality, not target to
create a search_space works any place else.

1. add __post_init__ to load data from deserialized data.
2. unfortunately, #1 doesn't work automatically like document. So add
decode_set_space to make it can be called explictly.
3. override add, update methods, because they are used in code.
4. call decoders in schema.

squirrelsc committed:
Improve os requirements support super types

for example, Unbuntu satisifies Linux

Co-authored-by: Leandro A. F. Pereira <leandro@hardinfo.org>

squirrelsc committed:
Improve ssh shell to support reconnect

It needs to call close() explictly. In future, it may need reconnect on
exception automatically.

squirrelsc committed:
Add Feature for plugable platform functionality

Feature uses to add platform specified features. some scenarios,
1. Almost platform support start and stop an VM, but very different
   behavior. Each platform can implement customized start/stop method.
2. Some feature is enabled on special vms or os, it needs customzied
   code to skip unsupported environment. like dpdk is supported by 5.4
   or higher kernels.

changes,
1. Add Feature class. Feature is base class of all features, Feature
    requirement are checked on node level.
2. Add Features class. Features is the feature collection on a node. It
    uses to initialize a feature, if it's requested by test case.
3. Add supported_features on platform. Platform must indicates which
    types are used to create a feature.

squirrelsc committed:
Support features on simple_requirement

squirrelsc committed:
refactor azure code for adding feature.

move shared code out of azure.py

squirrelsc committed:
Add StartStop feature and azure implementation.

For convenient, it provides start, stop and restart methods.

squirrelsc committed:
Add restart test case for demostration

squirrelsc committed:
improve ssh connection stability

The connection is observed failed on exec_command after rebooting a node.
So retry more steps to make it more stable.

Also adjust the order of code to group them by meaning.

squirrelsc committed:
fix failure on withscript case.

The timer is not reliable on local run.

squirrelsc committed:
Remove config method in most places.

In current stage, the design is clearer. In most cases, it doesn't need
a separated config method to pass in parameters in different stages.
config can be passed in when calling __init__ or _initialize.
The config in platform is kept, as it needs more refactoring, may
refactor it later.

squirrelsc committed:
Support parameters in InitializableMixin

In some cases, parameters is decided when initialization, add parameter
to support it. It will be used in future refactoring.

squirrelsc committed:
remove test_runner folder

There is no plan to integrate other runners more than lisarunner,
so remove the folder.

squirrelsc committed:
remove hardcoded disgnostics profile

It uses to support serial console log, but it needs a precreated storage.
There is another way called managed storage. Will investigate how to
upport it later.

squirrelsc committed:
add missed () of path.exist()

squirrelsc committed:
Improve process to save result.

if a process result is requested twice, it can return immediately.

squirrelsc committed:
azure: improve error message

load lowest error details. The error may be an object, sometime the
message is a serialized dict. So try to deserialize and parse it.

squirrelsc committed:
Support more operating systems.

squirrelsc committed:
Don't skip test cases when not to create an env

Even test cases is not allowed to create an env, it may match another
env in runtime. So remove a skip.

squirrelsc committed:
environment: remove capability cache

Capability is not queried frequently, so remove the cache.
So it doesn't need to manage (reset) the cache anymore.

squirrelsc committed:
add dmesg tool

Co-authored-by: Leandro A. F. Pereira <leandro@hardinfo.org>

squirrelsc committed:
add filename pattern secret

for example,
helloworld.yml will show like h***d.yml

squirrelsc committed:
support log in tool

squirrelsc committed:
small improvement on comments and log

squirrelsc committed:
fix runbook env cannot be created

allow_create should apply to case requirements only.

squirrelsc committed:
real feature sets should be copied in min value

The capability won't be changed when generate min value. So just copy
it.

squirrelsc committed:
change provisioning case to a real case from demo

squirrelsc committed:
add azure runbooks

The runbook are official things, not for demo purpose. Once user fills
in secrets, the cases should cover specified scope.

squirrelsc committed:
azure: support tags on availability set.

lpereira reviewed (dismissed):
:shipit:

squirrelsc committed:
shell, add log and make it try quicker.

squirrelsc committed:
arm-template Remove pass-through variables

Use parameter in template directly.

squirrelsc committed:
Add availabilityset properties support

squirrelsc committed:
add metadata for future use

squirrelsc committed:
azure: fix location are not used actually

As current design all resource should be in same location, so
1. use first node's location for resources except vm.
2. vm uses its own location, though it's the same currently.

squirrelsc committed:
remove anti-pattern path exists check before open

squirrelsc committed:
Add subclasses factory

The major logic is from platform_, and will replace it.
1. Factory supports to create an object form an string.
2. create_by_type_name, given a type name, and return an object.
3. create_by_runbook, given a runbook with type field. return an object.

squirrelsc committed:
refactor platform to use subclasses factory

remove abstractmethod, as abstract class cannot be used as type
parameter.

squirrelsc committed:
exit with non-zero code, if any case failed.

squirrelsc committed:
fix failed hello world case.

the _name should be set earlier.

squirrelsc committed:
implement notifier

messages can be saved by notifiers.

squirrelsc committed:
notifier support extendable schema

squirrelsc committed:
integrate notifier to main flow.

squirrelsc committed:
add test result message, and fix direct set status

rename env to be short and clear

squirrelsc committed:
add action message

squirrelsc committed:
implement sample console notifier

andyleejordan reviewed (commented):
Mostly I have questions 😄

squirrelsc committed:
add comments on InitializableMixin and feature.

squirrelsc committed:
Fix platform ut to match real behavior

The capability should be a fixed number, not space in prepare phase.
Fix related test case expectation.

andyleejordan reviewed (approved):
There's fixes to be made but I think I'd rather see this merged sooner rather than later, as I have more work on top of it. What do you think @lpereira?
